sdk/go/pkg/state/client.go

```diff
@@ -0,0 +1,284 @@
+	WithEnvScope() ScopedState
+	WithCustomScope(scope StateScope) ScopedState
+}
+
+type boundStateBuilder struct {
+	httpClient     *http.Client
+	baseURL        string
+	token          string
+	organizationID string
+	projectID      string
+	environmentID  string
+}
+
+func NewBoundState(req sdk.Request) StateScopeBuilder {
+	return &boundStateBuilder{
+		httpClient:     httputil.NewRetriableHTTPClient().StandardClient(),
+		baseURL:        req.MetaString("stateStoreUrl"),
+		token:          req.MetaString("stateStoreToken"),
+		organizationID: req.MetaString("organizationID"),
+		projectID:      req.MetaString("projectID"),
+		environmentID:  req.MetaString("environmentID"),
+	}
+}
+
+func (b *boundStateBuilder) WithOrgScope() ScopedState {
+	return b.WithCustomScope(StateScope{OrganizationID: b.organizationID})
+}
+
+func (b *boundStateBuilder) WithProjectScope() ScopedState {
+	return b.WithCustomScope(StateScope{
+		OrganizationID: b.organizationID,
+		ProjectID:      &b.projectID,
+	})
+}
+
+func (b *boundStateBuilder) WithEnvScope() ScopedState {
+	return b.WithCustomScope(StateScope{
+		OrganizationID: b.organizationID,
+		ProjectID:      &b.projectID,
+		EnvironmentID:  &b.environmentID,
+	})
+}
+
+func (b *boundStateBuilder) WithCustomScope(scope StateScope) ScopedState {
+	return &boundState{
+		httpClient:     b.httpClient,
+		baseURL:        b.baseURL,
+		token:          b.token,
+		organizationID: b.organizationID,
+		projectID:      b.projectID,
+		environmentID:  b.environmentID,
+		scope:          scope,
+	}
+}
+
+type ScopedState interface {
+	Get(ctx context.Context, key string) (json.RawMessage, uint32, error)
+	Set(ctx context.Context, key string, updateFn func(old json.RawMessage) (json.RawMessage, error)) error
+	SetKV(ctx context.Context, key string, value json.RawMessage, version uint32) error
+	Delete(ctx context.Context, key string) error
+}
+
+type boundState struct {
+	httpClient     *http.Client
+	baseURL        string
+	token          string
+	organizationID string
+	projectID      string
+	environmentID  string
+	scope          StateScope
+}
+
+type stateResponse struct {
+	Value   json.RawMessage `json:"value"`
+	Version uint32          `json:"version"`
+}
+
+func (b *boundState) injectHeaders(req *http.Request) {
+	req.Header.Set("X-Eaas-State-Token", b.token)
+	req.Header.Set("X-Organization-ID", b.organizationID)
+	req.Header.Set("X-Project-ID", b.projectID)
+	req.Header.Set("X-Environment-ID", b.environmentID)
+	req.Header.Set("Content-Type", "application/json")
+}
+
+func (b *boundState) Get(ctx context.Context, key string) (json.RawMessage, uint32, error) {
+	var zero json.RawMessage
+	value, version, err := b.getRaw(ctx, key)
+	if err != nil {
+		return zero, 0, sdk.NewErrNotFound(fmt.Sprintf("key not found: %s", key))
+	}
+	return value, version, nil
+}
+
+// Developer passes key and value to set
+func (b *boundState) SetKV(ctx context.Context, key string, value json.RawMessage, version uint32) error {
```

Comment:
Set's the value for given key and scope, OCC needs to be handled by function developer